frontend/public/components/monitoring/receiver-forms/alert-manager-receiver-forms.tsx

```diff
@@ -0,0 +1,406 @@
+ * }
+ */
+const createRoute = (
+  receiver: AlertmanagerReceiver,
+  routeLabels: RouteEditorLabel[],
+): AlertmanagerRoute => {
+  return _.reduce(
+    routeLabels,
+    (acc, label) => {
+      _.set(acc, [label.isRegex ? 'match_re' : 'match', label.name], label.value);
+      return acc;
+    },
+    { receiver: receiver.name },
+  );
+};
+
+/**
+ * Returns new Receiver object
+ * Ex:
+ * {
+ *   name: MyNewReceiver
+ *   pagerduty_configs: {
+ *     routing_key: <integration_key>
+ *   }
+ * }
+ */
+const createReceiver = (
+  formValues,
+  createReceiverConfig: Function,
+  receiverToEdit: AlertmanagerReceiver,
+): AlertmanagerReceiver => {
+  const receiverConfig = createReceiverConfig(
+    formValues,
+    receiverToEdit && receiverToEdit[formValues.receiverType]
+      ? receiverToEdit[formValues.receiverType][0] // pass in receiver config if editing existing receiver
+      : {},
+  );
+  return {
+    name: formValues.receiverName,
+    [formValues.receiverType]: [{ ...receiverConfig }],
+  };
+};
+
+const subformFactory = (receiverType: string) => {
+  switch (receiverType) {
+    case 'pagerduty_configs':
+      return PagerDutyForm;
+    case 'webhook_configs':
+      return WebhookForm;
+    default:
+      return;
+  }
+};
+
+const useFormStore = (initialState) => {
+  const [formValues, setFormValues] = React.useState(initialState);
+
+  const handleChange = (event) => {
+    setFormValues({
+      ...formValues,
+      [event.target.name]: event.target.value,
+    });
+  };
+
+  return {
+    formValues,
+    setFormValues,
+    handleChange,
+  };
+};
+
+const getRouteLabelsForEditor = (
+  isDefaultReceiver: boolean,
+  receiverToEdit: AlertmanagerReceiver,
+  allRoutes: AlertmanagerRoute[],
+): RouteEditorLabel[] => {
+  const routeLabels = convertReceiverRoutesToEditorLabels(receiverToEdit, allRoutes);
+  return !isDefaultReceiver && _.isEmpty(routeLabels)
+    ? [{ name: '', value: '', isRegex: false }]
+    : routeLabels;
+};
+
+const ReceiverBaseForm: React.FC<ReceiverBaseFormProps> = ({
+  obj: secret, // Secret "alertmanager-main" which contains alertmanager.yaml config
+  titleVerb,
+  saveButtonText,
+  editReceiverNamed,
+}) => {
+  const [errorMsg, setErrorMsg] = React.useState<string>();
+  const [inProgress, setInProgress] = React.useState<boolean>(false);
+  let config: AlertmanagerConfig;
+  if (!errorMsg) {
+    config = getAlertmanagerConfig(secret, setErrorMsg);
+  }
+
+  const INITIAL_STATE = {
+    receiverName: '',
+    receiverType: '',
+    routeLabels: [],
+    ...PagerDutyForm.getInitialValues(null),
+    ...WebhookForm.getInitialValues(null),
+  };
+
+  let receiverToEdit: AlertmanagerReceiver;
+  if (editReceiverNamed) {
+    receiverToEdit = _.find(_.get(config, 'receivers'), { name: editReceiverNamed });
+    INITIAL_STATE.receiverName = receiverToEdit ? receiverToEdit.name : '';
+    INITIAL_STATE.receiverType = _.find(_.keys(receiverToEdit), (key) =>
+      _.includes(key, '_configs'),
```

Comment:
Should this be `_.endsWith` instead of `_.includes`?

Reply:
yes, fixed